### PR TITLE
Allow multiple (intermediate) CA certs

### PR DESCRIPTION
**Pull request by knoppiks**
This PR tries to enable the dynamiclistener to hand out not only the first signing certificate it finds in the CA chain.

I stumbled on [this discussion](https://github.com/k3s-io/k3s/discussions/7596) while searching for the exact problem in k3s, where @brandond pointed to this repository.

Maybe it would be good to add tests but being not a go expert I struggle to comprehend the tests conducted on the listener.

### CODE REVIEW

**Comment by brandond** [comment]
This looks good structurally, however I don't think we can change the call signatures of existing exported functions in order to avoid breaking users of this library. We'd probably need to modify this PR to include new functions that return `[]x509.Certificate`, and having those be used to load the cert, instead of simply changing the existing ones.

**Comment by knoppiks** [comment]
@brandond Please have a look. I tried my best with the naming, maybe you have better suggestions.

File: factory/ca.go
Comment (brandond) [inline factory/ca.go:72]: You'll get a panic here if `LoadCAChain` returns a non-nil error, as you can't do `chain[0]` on an empty array.
```suggestion
	if err != nil {
		return nil, nil, err
	}
	return chain[0], signer, nil
```

File: factory/ca.go
Comment (brandond) [inline factory/ca.go:96]: same issue as above with `chain[0]`

File: listener.go
Comment (brandond) [inline listener.go:42]: Give this a more descriptive name

```suggestion
func NewListenerWithChain(l net.Listener, storage TLSStorage, caCert []*x509.Certificate, caKey crypto.Signer, config Config) (net.Listener, http.Handler, error) {
```

**Review by brandond — CHANGES_REQUESTED** [review]
couple nits on error handing and naming, looks good otherwise!

File: factory/ca.go
Comment (knoppiks) [reply factory/ca.go:105]: Very good hint, thank you. Found two more places with the same problem.

**Review by brandond — APPROVED** [review]

**Comment by knoppiks** [comment]
Any news here? I'd really like to tackle the k3s portion of this fix.

**Comment by caroline-suse-rancher** [comment]
Hey @knoppiks we're in code freeze at the moment for July releases, but after that we'll be looking for one more reviewer/approver for this PR. I will bring it up with the team after freeze is lifted!

**Comment by brandond** [comment]
We can get it merged here, and then hold off on updating anything in K3s until after the freeze is over. This repo is not subject to code freeze.

**Review by briandowns — APPROVED** [review]

**Review by dereknola — APPROVED** [review]

**Pull request merged**
